src/ProjectTask.php

```diff
@@ -555,6 +561,59 @@ public function post_clone($source, $history)
         }
     }
 
+    /**
+     * Set automatically the real start and end dates if not set
+     * @param array $input // the input of the form
+     * @param array $fields // the old fields of the task
+     *
+     * @return array // the input with the real start and end dates set if needed
+     */
+    public static function autoSetDate(array $input, array $fields): array
+    {
+        if (!empty($input['real_start_date']) && !empty($input['real_end_date']) || empty($input['percent_done'])) {
+            return $input;
+        }
+        if (isset($fields['id']) && !empty($fields['id'])) {
+            if (empty($fields['real_start_date']) && ($input['percent_done'] > 0)) {
+                $input['real_start_date'] = \Session::getCurrentTime();
+            }
+            if (empty($fields['real_end_date']) && $input['percent_done'] == 100) {
+                $input['real_end_date'] = \Session::getCurrentTime();
+            }
+        } else {
+            if ($input['percent_done'] == 100) {
+                $input['real_end_date'] = \Session::getCurrentTime();
+            }
+            if ($input['percent_done'] > 0) {
+                $input['real_start_date'] = \Session::getCurrentTime();
+            }
+        }
```

Comment:
```suggestion
        $real_start_date =  $input['real_start_date'] ?: $this->fields['real_start_date'] ?: null;
        $real_end_date =  $input['real_end_date'] ?: $this->fields['real_end_date'] ?: null;
        
        if ($real_start_date === null && ($input['percent_done'] > 0)) {
            $input['real_start_date'] = \Session::getCurrentTime();  
        } 
        if ($real_end_date === null && ($input['percent_done'] == 100)) {
            $input['real_end_date'] = \Session::getCurrentTime();
        }
```

Small suggestion to avoid nested `if` (better readability).